README.md
## Docs

[Development](docs/DEVELOPMENT.md)

[Deployment](docs/DEPLOYMENT.md)

[History](docs/HISTORY.md)

## Security

[Security Policy](SECURITY.md)

## License

[MIT](LICENSE.txt)


docs/HISTORY.md
# HISTORY (How it happened)

## Early

1. PogoToken Deployed
2. MasterChef Deployed
3. PogoToken Ownership Transfered to MasterChef
4. TimeLock Deployed
5. PogoFactory Deployed
6. PogoRouter Deployed
7. Migrator Deployed

## Pre-migration

Incentivised pools for Uniswap LP tokens were added by Master Chef to encourage locking liquidity into the platform before eventual migration of those Uniswap LP tokens to PogoSwap SLP tokens.

## Migration

...

## Post-migration

...


tasks/migrate.js
const { getApprovalDigest } = require("./utilities")

const { ecsign } = require("ethereumjs-util")

module.exports = async function (
  { tokenA, tokenB },
  { 
    getChainId, 
    ethers: {
      getNamedSigner,
      utils: { hexlify },
      constants: { MaxUint256 },
      Wallet,
    }
  },
  runSuper
) {

  console.log("Migrate", config.networks[hre.network.name].accounts)

  // Dev private key
  const privateKey = Wallet.fromMnemonic(config.networks[hre.network.name].accounts.mnemonic, "m/44'/60'/0'/0/1").privateKey

  const erc20Contract = await ethers.getContractFactory("UniswapV2ERC20")

  const token = erc20Contract.attach("0x1c5DEe94a34D795f9EEeF830B68B80e44868d316")

  const deadline = MaxUint256

  const dev = await getNamedSigner("dev")

  const nonce = await token.connect(dev).nonces(dev.address)

  const PogoRoll = await ethers.getContract("PogoRoll")

  const chainId = await getChainId()

  const digest = await getApprovalDigest(
    token,
    chainId,
    {
      owner: dev.address,
      spender: PogoRoll.address,
      value: await token.balanceOf(dev.address),
    },
    nonce,
    deadline
  )

  const { v, r, s } = ecsign(
    Buffer.from(digest.slice(2), "hex"),
    Buffer.from(privateKey.slice(2), "hex")
  )

  console.log({ v, r: hexlify(r), s: hexlify(s) })

  const migrateTx = await PogoRoll
    .connect(dev)
    .migrateWithPermit(
      tokenA,
      tokenB,
      await token.balanceOf(dev.address),
      0,
      0,
      deadline,
      v,
      hexlify(r),
      hexlify(s),
      {
        gasLimit: 8000000,
        gasPrice: 100000000000,
      }
    )

  await migrateTx.wait()

  console.log(migrateTx)
}


hardhat.config.ts
import "dotenv/config"
import "@nomiclabs/hardhat-etherscan"
import "@nomiclabs/hardhat-solhint"
import "@tenderly/hardhat-tenderly"
import "@nomiclabs/hardhat-waffle"
import "hardhat-abi-exporter"
import "hardhat-deploy"
import "hardhat-deploy-ethers"
import "hardhat-gas-reporter"
import "hardhat-spdx-license-identifier"
import "hardhat-typechain"
import "hardhat-watcher"
import "solidity-coverage"
import "./tasks"

import { HardhatUserConfig } from "hardhat/types"
import { removeConsoleLog } from "hardhat-preprocessor"

const API_URL = "";
const PRIV_KEY = "";
const PUBLIC_KEY = "";
const DEPLOYER_ACCOUNT = "";
const DEV_ACCOUNT = "";
const PRIVATE_KEY = "";
const POLYSCAN_API_KEY = "";
const FORKING = false;
const HARDHAT_MAX_MEMORY = 4096;
const HARDHAT_SHOW_STACK_TRACES = true;
const HARDHAT_VERBOSE = true;
const REPORT_GAS = false;

// const accounts = {
//   mnemonic: process.env.MNEMONIC || "test test test test test test test test test test test junk",
//   // accountsBalance: "990000000000000000000",
// }

const config: HardhatUserConfig = {
	abiExporter: {
    path: "./abi",
    clear: false,
    flat: true,
    // only: [],
    // except: []
  },
	defaultNetwork: "hardhat",
	etherscan: {
    apiKey: `${POLYSCAN_API_KEY}`,
  },

  mocha: {
    timeout: 20000,
  },
  // CHIGAG 1: need to add deployer and dev accounts to list below for each network!!!
  namedAccounts: {
    deployer: {
      default: 0,
      80001: `${DEPLOYER_ACCOUNT}`,
      97: `${DEPLOYER_ACCOUNT}`,

    },
    dev: {
      // Default to 1
      default: 1,
      80001: `${DEV_ACCOUNT}`,
      97: `${DEV_ACCOUNT}`,
      // dev address mainnet
      // 1: "",
    },
  },
  networks: {
		localhost: {
      url: 'http://localhost:8545/rpc',
      timeout: 600000,
      gasPrice: 900000000,
      gas: 21000000,
      accounts: [ '0xdf57089febbacf7ba0bc227dafbffa9fc08a93fdc68e1e42411a14efcf23656e' ],
		},
		// mainnet: {
    //   url: `https://mainnet.infura.io/v3/${process.env.INFURA_API_KEY}`,
    //   accounts: process.env.MNEMONIC,
    //   gasPrice: 120 * 1000000000,
    //   chainId: 1,
    // },
    // hardhat: {
    //   forking: {
    //     enabled: process.env.FORKING === "true",
    //     url: `https://eth-mainnet.alchemyapi.io/v2/${process.env.ALCHEMY_API_KEY}`,
    //   },
    //   live: false,
    //   saveDeployments: true,
    //   tags: ["test", "local"],
    // },
    // ropsten: {
    //   url: `https://ropsten.infura.io/v3/${process.env.INFURA_API_KEY}`,
    //   accounts,
    //   chainId: 3,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasPrice: 5000000000,
    //   gasMultiplier: 2,
    // },
    // rinkeby: {
    //   url: `https://rinkeby.infura.io/v3/${process.env.INFURA_API_KEY}`,
    //   accounts,
    //   chainId: 4,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasPrice: 5000000000,
    //   gasMultiplier: 2,
    // },
    // goerli: {
    //   url: `https://goerli.infura.io/v3/${process.env.INFURA_API_KEY}`,
    //   accounts,
    //   chainId: 5,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasPrice: 5000000000,
    //   gasMultiplier: 2,
    // },
    // kovan: {
    //   url: `https://kovan.infura.io/v3/${process.env.INFURA_API_KEY}`,
    //   accounts,
    //   chainId: 42,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasPrice: 20000000000,
    //   gasMultiplier: 2,
    // },
    // moonbase: {
    //   url: "https://rpc.testnet.moonbeam.network",
    //   accounts,
    //   chainId: 1287,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gas: 5198000,
    //   gasMultiplier: 4,
    // },
    // fantom: {
    //   url: "https://rpcapi.fantom.network",
    //   accounts,
    //   chainId: 250,
    //   live: true,
    //   saveDeployments: true,
    //   gasPrice: 22000000000,
    // },
    // "fantom-testnet": {
    //   url: "https://rpc.testnet.fantom.network",
    //   accounts,
    //   chainId: 4002,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasMultiplier: 2,
    // },
     matic: {
       url: `${API_URL}`,
       accounts: [PRIV_KEY],
       chainId: 137,
       live: true,
       saveDeployments: true,
     },
     mumbai: {
       url: `${API_URL}`,
       accounts: [PRIV_KEY],
       chainId: 80001,
			 live: true,
       saveDeployments: true,
    },
    // xdai: {
    //   url: "https://rpc.xdaichain.com",
    //   accounts,
    //   chainId: 100,
    //   live: true,
    //   saveDeployments: true,
    // },
   /* bsc: {
      url: "https://bsc-dataseed.binance.org",
      accounts: [process.env.MNEMONIC],
      chainId: 56,
      live: true,
      saveDeployments: true,
    },
    "bsc-testnet": {
      url: "https://data-seed-prebsc-2-s3.binance.org:8545",
      accounts: [process.env.MNEMONIC],
      chainId: 97,
      live: true,
      saveDeployments: true,
      tags: ["staging"],
      gasMultiplier: 2,
    },*/
    // heco: {
    //   url: "https://http-mainnet.hecochain.com",
    //   accounts,
    //   chainId: 128,
    //   live: true,
    //   saveDeployments: true,
    // },
    // "heco-testnet": {
    //   url: "https://http-testnet.hecochain.com",
    //   accounts,
    //   chainId: 256,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasMultiplier: 2,
    // },
    // avalanche: {
    //   url: "https://api.avax.network/ext/bc/C/rpc",
    //   accounts,
    //   chainId: 43114,
    //   live: true,
    //   saveDeployments: true,
    //   gasPrice: 470000000000,
    // },
    // fuji: {
    //   url: "https://api.avax-test.network/ext/bc/C/rpc",
    //   accounts,
    //   chainId: 43113,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasMultiplier: 2,
    // },
    // harmony: {
    //   url: "https://api.s0.t.hmny.io",
    //   accounts,
    //   chainId: 1666600000,
    //   live: true,
    //   saveDeployments: true,
    //   gasMultiplier: 2,
    // },
    // "harmony-testnet": {
    //   url: "https://api.s0.b.hmny.io",
    //   accounts,
    //   chainId: 1666700000,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasMultiplier: 2,
    // },
    // okex: {
    //   url: "https://exchainrpc.okex.org",
    //   accounts,
    //   chainId: 66,
    //   live: true,
    //   saveDeployments: true,
    // },
    // "okex-testnet": {
    //   url: "https://exchaintestrpc.okex.org",
    //   accounts,
    //   chainId: 65,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasMultiplier: 2,
    // },
    // arbitrum: {
    //   url: "https://arb1.arbitrum.io/rpc",
    //   accounts,
    //   chainId: 42161,
    //   live: true,
    //   saveDeployments: true,
    //   blockGasLimit: 700000,
    // },
    // "arbitrum-testnet": {
    //   url: "https://kovan3.arbitrum.io/rpc",
    //   accounts,
    //   chainId: 79377087078960,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasMultiplier: 2,
    // },
    // celo: {
    //   url: "https://forno.celo.org",
    //   accounts,
    //   chainId: 42220,
    //   live: true,
    //   saveDeployments: true,
    // },
    // palm: {
    //   url: "https://palm-mainnet.infura.io/v3/da5fbfafcca14b109e2665290681e267",
    //   accounts,
    //   chainId: 11297108109,
    //   live: true,
    //   saveDeployments: true,
    // },
    // "palm-testnet": {
    //   url: "https://palm-testnet.infura.io/v3/da5fbfafcca14b109e2665290681e267",
    //   accounts,
    //   chainId: 11297108099,
    //   live: true,
    //   saveDeployments: true,
    //   tags: ["staging"],
    //   gasMultiplier: 2,
    // },
    // moonriver: {
    //   url: "https://rpc.moonriver.moonbeam.network",
    //   accounts,
    //   chainId: 1285,
    //   live: true,
    //   saveDeployments: true,
    // },
    // fuse: {
    //   url: "https://rpc.fuse.io",
    //   accounts,
    //   chainId: 122,
    //   live: true,
    //   saveDeployments: true,
    // },
    // clover: {
    //   url: "https://rpc-ivy.clover.finance",
    //   accounts,
    //   chainId: 1024,
    //   live: true,
    //   saveDeployments: true,
    // },
    // telos: {
    //   url: "https://rpc1.us.telos.net/evm",
    //   accounts,
    //   chainId: 40,
    //   live: true,
    //   saveDeployments: true,
    // },
    // moonbeam: {
    //   url: "https://rpc.api.moonbeam.network",
    //   accounts,
    //   chainId: 1284,
    //   live: true,
    //   saveDeployments: true,
    // }, 
	},
  paths: {
    artifacts: "artifacts",
    cache: "cache",
    deploy: "deploy",
    deployments: "deployments",
    imports: "imports",
    sources: "contracts",
    tests: "test",
  },
  preprocess: {
    eachLine: removeConsoleLog((bre) => bre.network.name !== "hardhat" && bre.network.name !== "localhost"),
  },
	solidity: {
    compilers: [
      {
        version: "0.6.12",
        settings: {
          optimizer: {
            enabled: true,
            runs: 200,
          },
        },
      },
      {
        version: "0.8.13",
        settings: {
          optimizer: {
            enabled: true,
            runs: 1000,
          },
        },
      },
    ],
  },

  spdxLicenseIdentifier: {
    overwrite: false,
    runOnCompile: true,
  },
  tenderly: {
    project: process.env.TENDERLY_PROJECT!,
    username: process.env.TENDERLY_USERNAME!,
  },
  typechain: {
    outDir: "types",
    target: "ethers-v5",
  },
  watcher: {
    compile: {
      tasks: ["compile"],
      files: ["./contracts"],
      verbose: true,
    },
  },
}

export default config


deploy/MasterChef.js
module.exports = async function ({ ethers, deployments, getNamedAccounts }) {
  const { deploy } = deployments

  const { deployer } = await getNamedAccounts()

  const PogoTokenAddress = "0x303D22b4cd2A5c811aD95f13787E7f1E71BE1899"
  
  const Pogo = await ethers.getContract("PogoToken")
  
  const { address } = await deploy("MasterChefV2", {
    from: deployer,
    args: [],
    log: true,
    deterministicDeployment: false
  })

//  if (await Pogo.owner() !== address) {
    // Transfer Pogo Ownership to Chef
//    console.log("Transfer Pogo Ownership to Chef")
//    await (await Pogo.transferOwnership(address)).wait()
//  }

//	const masterChef = await ethers.getContract("MasterChefV2")
//  if (await masterChef.owner() !== deployer) {
    // Transfer ownership of MasterChef to dev
  //  console.log("Transfer ownership of MasterChefV2 to dev")
	//    await (await masterChef.transferOwnership(dev)).wait()
//  }
},

module.exports.tags = ["MasterChefV2"]
module.exports.dependencies = ["UniswapV2Factory", "UniswapV2Router02", "PogoToken"]


tasks/index.js
const { task } = require("hardhat/config");

const {
  ethers: {
    constants: { MaxUint256 },
  },
  utils: { defaultAbiCoder },
} = require("ethers");
const { MINICHEF_ADDRESS } = require("@sushiswap/core-sdk");

const fs = require("fs");

function getSortedFiles(dependenciesGraph) {
  const tsort = require("tsort");
  const graph = tsort();

  const filesMap = {};
  const resolvedFiles = dependenciesGraph.getResolvedFiles();
  resolvedFiles.forEach((f) => (filesMap[f.sourceName] = f));

  for (const [from, deps] of dependenciesGraph.entries()) {
    for (const to of deps) {
      graph.add(to.sourceName, from.sourceName);
    }
  }

  const topologicalSortedNames = graph.sort();

  // If an entry has no dependency it won't be included in the graph, so we
  // add them and then dedup the array
  const withEntries = topologicalSortedNames.concat(
    resolvedFiles.map((f) => f.sourceName)
  );

  const sortedNames = [...new Set(withEntries)];
  return sortedNames.map((n) => filesMap[n]);
}

function getFileWithoutImports(resolvedFile) {
  const IMPORT_SOLIDITY_REGEX = /^\s*import(\s+)[\s\S]*?;\s*$/gm;

  return resolvedFile.content.rawContent
    .replace(IMPORT_SOLIDITY_REGEX, "")
    .trim();
}

subtask(
  "flat:get-flattened-sources",
  "Returns all contracts and their dependencies flattened"
)
  .addOptionalParam("files", undefined, undefined, types.any)
  .addOptionalParam("output", undefined, undefined, types.string)
  .setAction(async ({ files, output }, { run }) => {
    const dependencyGraph = await run("flat:get-dependency-graph", { files });
    console.log(dependencyGraph);

    let flattened = "";

    if (dependencyGraph.getResolvedFiles().length === 0) {
      return flattened;
    }

    const sortedFiles = getSortedFiles(dependencyGraph);

    let isFirst = true;
    for (const file of sortedFiles) {
      if (!isFirst) {
        flattened += "\n";
      }
      flattened += `// File ${file.getVersionedName()}\n`;
      flattened += `${getFileWithoutImports(file)}\n`;

      isFirst = false;
    }

    // Remove every line started with "// SPDX-License-Identifier:"
    flattened = flattened.replace(
      /SPDX-License-Identifier:/gm,
      "License-Identifier:"
    );

    flattened = `// SPDX-License-Identifier: MIXED\n\n${flattened}`;

    // Remove every line started with "pragma experimental ABIEncoderV2;" except the first one
    flattened = flattened.replace(
      /pragma experimental ABIEncoderV2;\n/gm,
      (
        (i) => (m) =>
          !i++ ? m : ""
      )(0)
    );

    flattened = flattened.trim();
    if (output) {
      console.log("Writing to", output);
      fs.writeFileSync(output, flattened);
      return "";
    }
    return flattened;
  });

subtask("flat:get-dependency-graph")
  .addOptionalParam("files", undefined, undefined, types.any)
  .setAction(async ({ files }, { run }) => {
    const sourcePaths =
      files === undefined
        ? await run("compile:solidity:get-source-paths")
        : files.map((f) => fs.realpathSync(f));

    const sourceNames = await run("compile:solidity:get-source-names", {
      sourcePaths,
    });

    const dependencyGraph = await run("compile:solidity:get-dependency-graph", {
      sourceNames,
    });

    return dependencyGraph;
  });

task("flat", "Flattens and prints contracts and their dependencies")
  .addOptionalVariadicPositionalParam(
    "files",
    "The files to flatten",
    undefined,
    types.inputFile
  )
  .addOptionalParam(
    "output",
    "Specify the output file",
    undefined,
    types.string
  )
  .setAction(async ({ files, output }, { run }) => {
    console.log(
      await run("flat:get-flattened-sources", {
        files,
        output,
      })
    );
  });

task("accounts", "Prints the list of accounts", require("./accounts"));
task("gas-price", "Prints gas price").setAction(async function (
  { address },
  { ethers }
) {
  console.log("Gas price", (await ethers.provider.getGasPrice()).toString());
});

task("bytecode", "Prints bytecode").setAction(async function (
  { address },
  { ethers }
) {
  console.log("Bytecode", await ethers.provider.getCode(address));
});

task("feeder:feed", "Feed").setAction(async function (
  { feedDev },
  { getNamedAccounts, ethers: { BigNumber }, getChainId }
) {
  const { deployer, dev } = await getNamedAccounts();

  const feeder = new ethers.Wallet(
    process.env.FEEDER_PRIVATE_KEY,
    ethers.provider
  );

  await (
    await feeder.sendTransaction({
      to: deployer,
      value: BigNumber.from(1).mul(BigNumber.from(10).pow(18)),
    })
  ).wait();
});

task("feeder:return", "Return funds to feeder").setAction(async function (
  { address },
  { ethers: { getNamedSigners } }
) {
  const { deployer, dev } = await getNamedSigners();

  await (
    await deployer.sendTransaction({
      to: process.env.FEEDER_PUBLIC_KEY,
      value: await deployer.getBalance(),
    })
  ).wait();

  await (
    await dev.sendTransaction({
      to: process.env.FEEDER_PUBLIC_KEY,
      value: await dev.getBalance(),
    })
  ).wait();
});

task("erc20:approve", "ERC20 approve")
  .addParam("token", "Token")
  .addParam("spender", "Spender")
  .addOptionalParam("deadline", MaxUint256)
  .setAction(async function (
    { token, spender, deadline },
    { ethers: { getNamedSigner } },
    runSuper
  ) {
    const erc20 = await ethers.getContractFactory("UniswapV2ERC20");

    const slp = erc20.attach(token);

    await (
      await slp.connect(await getNamedSigner("dev")).approve(spender, deadline)
    ).wait();
  });

task("factory:set-fee-to", "Factory set fee to")
  .addParam("feeTo", "Fee To")
  .setAction(async function (
    { feeTo },
    { ethers: { getNamedSigner } },
    runSuper
  ) {
    const factory = await ethers.getContract("UniswapV2Factory");
    console.log(`Setting factory feeTo to ${feeTo} address`);
    await (
      await factory.connect(await getNamedSigner("dev")).setFeeTo(feeTo)
    ).wait();
  });

task("factory:set-fee-to-setter", "Factory set fee to setter")
  .addParam("feeToSetter", "Fee To Setter")
  .setAction(async function (
    { feeToSetter },
    { ethers: { getNamedSigner } },
    runSuper
  ) {
    const factory = await ethers.getContract("UniswapV2Factory");
    console.log(`Setting factory fee to setter to ${feeToSetter} address`);
    await (
      await factory
        .connect(await getNamedSigner("dev"))
        .setFeeToSetter(feeToSetter)
    ).wait();
  });

// TODO: Swap?

// TODO: Test
task("router:add-liquidity", "Router add liquidity")
  .addParam("tokenA", "Token A")
  .addParam("tokenB", "Token B")
  .addParam("tokenADesired", "Token A Desired")
  .addParam("tokenBDesired", "Token B Desired")
  .addParam("tokenAMinimum", "Token A Minimum")
  .addParam("tokenBMinimum", "Token B Minimum")
  .addParam("to", "To")
  .addOptionalParam("deadline", MaxUint256)
  .setAction(async function (
    {
      tokenA,
      tokenB,
      tokenADesired,
      tokenBDesired,
      tokenAMinimum,
      tokenBMinimum,
      to,
      deadline,
    },
    { ethers: { getNamedSigner } },
    runSuper
  ) {
    const router = await ethers.getContract("UniswapV2Router");
    await run("erc20:approve", { token: tokenA, spender: router.address });
    await run("erc20:approve", { token: tokenB, spender: router.address });
    await (
      await router
        .connect(await getNamedSigner("dev"))
        .addLiquidity(
          tokenA,
          tokenB,
          tokenADesired,
          tokenBDesired,
          tokenAMinimum,
          tokenBMinimum,
          to,
          deadline
        )
    ).wait();
  });

// TODO: Test
task("router:add-liquidity-eth", "Router add liquidity eth")
  .addParam("token", "Token")
  .addParam("tokenDesired", "Token Desired")
  .addParam("tokenMinimum", "Token Minimum")
  .addParam("ethMinimum", "ETH Minimum")
  .addParam("to", "To")
  .addOptionalParam("deadline", MaxUint256)
  .setAction(async function (
    { token, tokenDesired, tokenMinimum, ethMinimum, to, deadline },
    { ethers: { getNamedSigner } },
    runSuper
  ) {
    const router = await ethers.getContract("UniswapV2Router");
    await run("erc20:approve", { token, spender: router.address });
    await (
      await router
        .connect(await getNamedSigner("dev"))
        .addLiquidityETH(
          token,
          tokenDesired,
          tokenMinimum,
          ethMinimum,
          to,
          deadline
        )
    ).wait();
  });

task("migrate", "Migrates liquidity from Uniswap to PogoSwap")
  .addOptionalParam(
    "a",
    "Token A",
    "0xaD6D458402F60fD3Bd25163575031ACDce07538D"
  )
  .addOptionalParam(
    "b",
    "Token B",
    "0xc778417E063141139Fce010982780140Aa0cD5Ab"
  )
  .setAction(require("./migrate"));

task("masterchef:add", "Add pool to masterchef").setAction(async function (
  taskArguments,
  { ethers: { getNamedSigner } },
  runSuper
) {
  const masterChef = await ethers.getContract("MasterChef");

  await (
    await masterChef
      .connect(await getNamedSigner("dev"))
      .add(1000, "0x3e78a806b127c02b54419191571d9379819e989c", true)
  ).wait();
});

task("masterchef:deposit", "MasterChef deposit")
  .addParam("pid", "Pool ID")
  .addParam("amount", "Amount")
  .setAction(async function (
    { pid, amount },
    { ethers: { getNamedSigner } },
    runSuper
  ) {
    const masterChef = await ethers.getContract("MasterChef");

    const { lpToken } = await masterChef.poolInfo(pid);

    await run("erc20:approve", { token: lpToken, spender: masterChef.address });

    await (
      await masterChef.connect(await getNamedSigner("dev")).deposit(pid, amount)
    ).wait();
  });

task("masterchef:withdraw", "MasterChef withdraw")
  .addParam("pid", "Pool ID")
  .addParam("amount", "Amount")
  .setAction(async function (
    { pid, amount },
    { ethers: { getNamedSigner } },
    runSuper
  ) {
    const masterChef = await ethers.getContract("MasterChef");

    const { lpToken } = await masterChef.poolInfo(pid);

    await run("erc20:approve", { token: lpToken, spender: masterChef.address });

    await (
      await masterChef
        .connect(await getNamedSigner("dev"))
        .withdraw(pid, amount)
    ).wait();
  });

task("bar:enter", "PogoBar enter")
  .addParam("amount", "Amount")
  .setAction(async function (
    { amount },
    { ethers: { getNamedSigner } },
    runSuper
  ) {
    const Pogo = await ethers.getContract("PogoToken");

    const bar = await ethers.getContract("PogoBar");

    await run("erc20:approve", { token: Pogo.address, spender: bar.address });

    await (await bar.connect(await getNamedSigner("dev")).enter(amount)).wait();
  });

task("bar:leave", "PogoBar leave")
  .addParam("amount", "Amount")
  .setAction(async function (
    { amount },
    { ethers: { getNamedSigner } },
    runSuper
  ) {
    const Pogo = await ethers.getContract("PogoToken");

    const bar = await ethers.getContract("PogoBar");

    await run("erc20:approve", { token: Pogo.address, spender: bar.address });

    await (await bar.connect(await getNamedSigner("dev")).leave(amount)).wait();
  });

task("maker:serve", "PogoBar serve")
  .addParam("a", "Token A")
  .addParam("b", "Token B")
  .setAction(async function (
    { a, b },
    { ethers: { getNamedSigner } },
    runSuper
  ) {
    const maker = await ethers.getContract("PogoMaker");

    await (
      await maker
        .connect(await getNamedSigner("dev"))
        .convert(a, b, { gasLimit: 5198000 })
    ).wait();
  });

task("deploy:complex-rewarder", "Deploy ComplexRewarder")
  .addParam("rewardToken", "Reward Token")
  .setAction(async function (
    { rewardToken },
    { ethers: { getNamedSigner }, getChainId, deployments },
    runSuper
  ) {
    const { deployer, dev } = await getNamedAccounts();
    const { deploy } = deployments;

    const chainId = await getChainId();

    let miniChefAddress;

    if (chainId === "31337") {
      miniChefAddress = (await deployments.get("MiniChefV2")).address;
    } else if (chainId in MINICHEF_ADDRESS) {
      miniChefAddress = MINICHEF_ADDRESS[chainId];
    } else {
      throw Error("No MINICHEF!");
    }

    const { address } = await deploy("ComplexRewarderTime", {
      from: deployer,
      args: [rewardToken, 0, miniChefAddress],
      log: true,
      deterministicDeployment: false,
    });

    console.log(`ComplexRewarderTime deployed at ${address}`);

    const complexRewarder = await ethers.getContract("ComplexRewarderTime");

    if ((await complexRewarder.owner()) !== dev) {
      console.log("Transfer ownership of ComplexRewarderTime to dev");
      await (await complexRewarder.transferOwnership(dev, true, false)).wait();
    }
  });

// task("deploy:clone-rewarder", "Deploy CloneRewarder")
// .addParam("rewardToken", "Reward Token")
// .addParam("lpToken", "LP Token")
// .addOptionalParam("rewardRate", "Reward Rate", 0)
// .setAction(async function ({ rewardToken, lpToken, rewardRate }, { getChainId, deployments }, runSuper) {
//   const { deployer, dev } = await getNamedAccounts();
//   const { deploy } = deployments;

//   const chainId = await getChainId();

//   let miniChefAddress;

//   if (chainId === "31337") {
//     miniChefAddress = (await deployments.get("MiniChefV2")).address;
//   } else if (chainId in MINICHEF_ADDRESS) {
//     miniChefAddress = MINICHEF_ADDRESS[chainId];
//   } else {
//     throw Error("No MINICHEF!");
//   }

//   const { address } = await deploy("CloneRewarderTime", {
//     from: deployer,
//     args: [miniChefAddress],
//     log: true,
//     deterministicDeployment: false,
//   });

//   console.log(`CloneRewarder deployed at ${address}`)

//   const cloneRewarder = await ethers.getContract("CloneRewarderTime");

//   const data = defaultAbiCoder.encode(['address', 'address', 'uint256', 'address'], [rewardToken, dev, rewardRate, lpToken])

//   await (await cloneRewarder.init(data)).wait()

//   if ((await complexRewarder.owner()) !== dev) {
//     console.log("Transfer ownership of CloneRewarderTime to dev");
//     await (await cloneRewarder.transferOwnership(dev, true, false)).wait();
//   }
// });
